recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+#!/usr/bin/env python3
+"""Recipe for pretraining wav2vec2. See config file for model definition.
+
+To run this recipe call python train.py hparams/train_it_wav2vec.yaml --find_unused_parameters --max_grad_norm 0.0
+"""
+
+import os
+import os.path as osp
+import re
+from tqdm.contrib import tqdm
+import logging
+import sys
+import time
+import tempfile
+import pathlib
+import functools
+from contextlib import ExitStack
+
+from types import SimpleNamespace
+import speechbrain as sb
+import torchaudio
+import torch
+import torch.nn.functional as F
+from torch.nn.parallel import DistributedDataParallel
+from torch.utils.data import DataLoader
+from hyperpyyaml import load_hyperpyyaml
+
+from speechbrain import Stage
+from speechbrain.utils.distributed import run_on_main
+from speechbrain.dataio.dataloader import SaveableDataLoader, LoopedLoader
+from speechbrain.dataio.sampler import DynamicBatchSampler, DistributedSamplerWrapper
+from speechbrain.lobes.models.wav2vec import make_padding_mask, compute_mask, W2VPaddedBatch
+
+
+PYTHON_VERSION_MAJOR = 3
+PYTHON_VERSION_MINOR = 7
+
+
+logger = logging.getLogger(__name__)
+
+
+@sb.utils.checkpoints.register_checkpoint_hooks
+class W2V2Brain(sb.core.Brain):
+    def compute_forward(self, batch, stage):
+        """Computes forward pass through wav2vec model and returns encoded and
+        target embeddings as well as other metrics of interest.
+        """
+        batch = batch.to(self.device)
+        wavs, wav_lens, padding_mask, mask = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+        padding_mask, mask = padding_mask.to(self.device), mask.to(self.device)
+
+        # normalisation already done in dataloader
+        results = self.modules.model(wavs, wav_lens, padding_mask, mask, is_pretraining=True,
```

Comment:
See, it might look nice because we only have one line. However, the purpose of compute_forward is to actually expose the forward computation. Again, if this is possible, maybe it would be great to decompose the wav2vec 2 forward here instead.

Reply:
For instance, we have absolutely no idea what are the outputs because of that.

Reply:
Yes after thinking about it I completely agree with you! The quantiser is not necessary for inference so it makes more sense to keep it seperate from the wav2vec object I think

Reply:
I'd like to discuss a few things about this.

1. I understand about the advantages of exposing the forward in compute_forward, but at the same time I like the idea that the pretrain forward logic is self-contained within the wav2vec2 class. Someone could simply import our SpeechBrain wav2vec2 module into another framework and easily do the pretrain forward with our "is_pretraining" flag. In any case, I see pros/cons in both approaches, what do you think?

2. Although the quantizer is not needed for inference, I'd suggest keeping it when we release our models, whether it is stored in the wav2vec object or not. I find that it is interesting to enable the community to play around with it, especially for research purposes (I have done explorations on HuggingFace's w2v2 quantizers myself).